stablehlo/dialect/StablehloOps.cpp

```diff
@@ -4164,6 +4164,71 @@ LogicalResult RngOp::reifyReturnTypeShapes(
 // SelectOp
 //===----------------------------------------------------------------------===//
 
+namespace {
+// Utility function, used by printSelectOpType and
+// printSameOperandsAndResultType
+ParseResult assignFromFunctionType(OpAsmParser& parser, llvm::SMLoc loc,
+                                   ArrayRef<Type*> operands, Type& result,
+                                   FunctionType& fnType) {
+  assert(fnType);
+  if (fnType.getInputs().size() != operands.size()) {
+    return parser.emitError(loc)
+           << operands.size() << " operands present, but expected "
+           << fnType.getInputs().size();
+  }
+  if (fnType.getResults().size() != 1) {
+    return parser.emitError(loc, "expected single output");
+  }
+
+  // Set operand types to function input types
+  for (auto [operand, input] : llvm::zip(operands, fnType.getInputs())) {
+    *operand = input;
+  }
+  result = fnType.getResults().front();
+  return success();
+}
+}  // namespace
+
+void printSelectOpType(OpAsmPrinter& p, Operation* op, Type pred, Type onTrue,
+                       Type onFalse, Type result) {
+  // Print functional type if true/false branches don't match return type.
+  if (onTrue != result || onFalse != result) {
+    p.printFunctionalType(op);
+    return;
+  }
+
+  // Print pred type and result type
+  p << pred << ", " << result;
+}
+
+ParseResult parseSelectOpType(OpAsmParser& parser, Type& pred, Type& onTrue,
+                              Type& onFalse, Type& result) {
+  // Operand and result types are the same, use copy constructor
+  llvm::SMLoc loc = parser.getCurrentLocation();
+  SmallVector<Type> types;
+  if (parser.parseTypeList(types)) {
+    return failure();
+  }
+
+  // Error handling for invalid types
+  // Fail if not two types, or single functional type
+  if (types.size() != 2 &&
+      (types.size() != 1 || !types.front().dyn_cast<FunctionType>())) {
+    return parser.emitError(loc,
+                            "expected functional type or list of two types");
+  }
+
+  if (types.size() == 2) {
+    pred = types.front();
+    onTrue = onFalse = result = types.back();
```

Comment:
Maybe it's just me but I find it easier to read `types[0]` and `types[1]` instead of `types.front()` and `types.back()`. The former is shorter, and the latter additionally doesn't require keeping track of the length of `types`. Similarly, for other parts of this PR.

Reply:
Only advantage is that front/back allows swapping of the underlying list type to one that doesn't implement indexing, i.e. `SmallVector -> std::list`. Generally preferred for polymorphic functions, but this function is not polymorphic, nor is `SmallVector` likely to change, so I see no harm in preferring the more readable `types[0]`.  

A note on the "doesn't require keeping track of the length of types", it still requires keeping track of the minimum length of size 2 to avoid undefined indexing OOB behavior, right?

Will update.